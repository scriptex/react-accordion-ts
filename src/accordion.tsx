import * as React from 'react';
import Panel from './panel';

interface AccordionItem {
	title: React.ReactNode;
	content: React.ReactNode;
}

interface AccordionState {
	activeTab: number;
}

interface Props {
	items: Array<AccordionItem>;
	duration: number;
}

class Accordion extends React.Component<Props> {
	public state: AccordionState = {
		activeTab: -1
	};

	constructor(props: Props) {
		super(props);
	}

	public activatePanel = (index: number): void => {
		this.setState((prev: AccordionState) => ({
			activeTab: prev.activeTab === index ? -1 : index
		}));
	}

	public render(): React.ReactNode {
		const { activeTab } = this.state;
		const { items, duration } = this.props;

		return (
			<div className="accordion" role="tablist">
				{Array.isArray(items) && items.length ? items.map(({ title, content }, index) => (
					<Panel
						key={index}
						title={title}
						index={index}
						duration={duration}
						activeTab={activeTab}
						activatePanel={this.activatePanel}
					>
						{content}
					</Panel>
				)) : ''}
			</div>
		);
	}
}

export default Accordion;
