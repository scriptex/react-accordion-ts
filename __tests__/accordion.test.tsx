import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { default as Accordion } from '../dist/accordion';

const mock = [
	{
		date: '12-10-2018',
		title: 'Awesome title',
		content: 'Fantastic content'
	},
	{
		date: '13-10-2018',
		title: 'Awesome title',
		content: 'Fantastic content'
	},
	{
		date: '13-10-2018',
		title: 'Awesome title',
		content: 'Fantastic content'
	}
];

describe('Accordion', () => {
	const items = mock.map(({ date, title, content }) => ({
		title: <h2>{date + ' - ' + title}</h2>,
		content: <p>{content}</p>
	}));

	it('should render properly', () => {
		const tree = renderer
			.create(<Accordion items={items} duration={300} multiple={true} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render properly', () => {
		const tree = renderer
			.create(<Accordion items={[]} duration={0} multiple={false} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render no items', () => {
		const tree = renderer
			.create(<Accordion items={undefined} duration={500} multiple={true} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
