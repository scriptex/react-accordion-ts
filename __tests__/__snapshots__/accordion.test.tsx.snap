// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render no items 1`] = `
<div
  className="accordion"
  role="tablist"
>
  
</div>
`;

exports[`Accordion should render properly 1`] = `
<div
  className="accordion"
  role="tablist"
>
  <div
    aria-expanded={false}
    className="panel"
    role="tabpanel"
  >
    <button
      className="panel__head"
      onClick={[Function]}
      role="tab"
    >
      <h2>
        12-10-2018 - Awesome title
      </h2>
    </button>
    <div
      aria-hidden={true}
      className="panel__body"
      style={
        Object {
          "height": "0px",
        }
      }
    >
      <div
        className="panel__content"
      >
        <p>
          Fantastic content
        </p>
      </div>
    </div>
  </div>
  <div
    aria-expanded={false}
    className="panel"
    role="tabpanel"
  >
    <button
      className="panel__head"
      onClick={[Function]}
      role="tab"
    >
      <h2>
        13-10-2018 - Awesome title
      </h2>
    </button>
    <div
      aria-hidden={true}
      className="panel__body"
      style={
        Object {
          "height": "0px",
        }
      }
    >
      <div
        className="panel__content"
      >
        <p>
          Fantastic content
        </p>
      </div>
    </div>
  </div>
  <div
    aria-expanded={false}
    className="panel"
    role="tabpanel"
  >
    <button
      className="panel__head"
      onClick={[Function]}
      role="tab"
    >
      <h2>
        13-10-2018 - Awesome title
      </h2>
    </button>
    <div
      aria-hidden={true}
      className="panel__body"
      style={
        Object {
          "height": "0px",
        }
      }
    >
      <div
        className="panel__content"
      >
        <p>
          Fantastic content
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion should render properly 2`] = `
<div
  className="accordion"
  role="tablist"
>
  
</div>
`;
